x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/user/ProfileHasPrivilegesRequest.java

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.core.security.action.user;
+
+import org.elasticsearch.action.ActionRequest;
+import org.elasticsearch.action.ActionRequestValidationException;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.xcontent.ConstructingObjectParser;
+import org.elasticsearch.xcontent.ObjectParser;
+import org.elasticsearch.xcontent.ParseField;
+import org.elasticsearch.xpack.core.security.authz.RoleDescriptor;
+import org.elasticsearch.xpack.core.security.authz.RoleDescriptor.IndicesPrivileges;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+
+import static org.elasticsearch.action.ValidateActions.addValidationError;
+import static org.elasticsearch.xcontent.ConstructingObjectParser.constructorArg;
+
+public class ProfileHasPrivilegesRequest extends ActionRequest {
+
+    @SuppressWarnings("unchecked")
+    public static final ConstructingObjectParser<ProfileHasPrivilegesRequest, Void> PARSER = new ConstructingObjectParser<>(
+        ProfileHasPrivilegesAction.NAME,
+        false,
+        arg -> new ProfileHasPrivilegesRequest((List<String>) arg[0], (RoleDescriptor) arg[1])
+    );
+    static {
+        PARSER.declareStringArray(constructorArg(), Fields.UIDS);
+        PARSER.declareField(
+            constructorArg(),
+            parser -> RoleDescriptor.parsePrivilegesCheck(ProfileHasPrivilegesAction.NAME, parser),
+            Fields.PRIVILEGES,
+            ObjectParser.ValueType.OBJECT
+        );
+    }
+
+    private List<String> uids;
+    private String[] clusterPrivileges;
+    private IndicesPrivileges[] indexPrivileges;
+    private RoleDescriptor.ApplicationResourcePrivileges[] applicationPrivileges;
+
+    public ProfileHasPrivilegesRequest(List<String> uids, RoleDescriptor privilegesToCheck) {
```

Comment:
I agree that the parsed `RoleDescriptor`, which is incomplete, needs to be validated before it is split into the index, cluster and application privileges.

Reply:
>Taking a step back, it is a concidence and conveinience that the privileges to check can re-use part of the RoleDescriptor code. But they are in fact two different things.

I think I haven't quite understood when I initially read your comment. 
I got that the constructor parameter should be `PrivilegesToCheck` instead of `RoleDescriptor`, but that the parsing from json should still return an incomplete `RoleDescriptor`.

I now get that we should avoid the `RoleDescriptor` class completely and instead use `PrivilegesToCheck`. I have pushed, which does that https://github.com/elastic/elasticsearch/pull/85898/commits/02c0b680ae07aba426cdfd6d53505305cb16aad4 .
Please let me know if this is what you were asking for, @ywangd .

Reply:
Ooops, also https://github.com/elastic/elasticsearch/pull/85898/commits/347de51ac53fedd43c6f0ee1211628ed31bd69b4

Reply:
Thanks @albertzaharovits 
Yes that's what I meant. Thanks for digging it.